### PR TITLE
Add ops(max_pool3d and max_pool3d_with_indices) | feat(atenlib)

### PR DESCRIPTION
**Pull request by fatcat-z**
1. Add max_pool3d and max_pool3d_with_indices ops.
2. Refactor the code so max_pool2d and max_pool3d could share the same helper function.
3. Refactor the code so max_pool2d_with_indices and max_pool3d_indices could share the same helper function. And move them to nn.py file.
4. Add tests.

### CODE REVIEW

File: onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py [hunk 1]
```diff
@@ -1006,6 +1011,28 @@ def _where_input_wrangler(
...
+        and sample.kwargs.get("padding") == 1,
+        reason="this combinations is not supported.",
+    ),
+    skip(
+        "nn.functional.max_pool3d",
+        matcher=lambda sample: sample.kwargs.get("return_indices") is True,
+        reason="this aten overload assume return_indices=False",
+    ),
+    skip(
+        "nn.functional.max_pool3d_with_indices",
```
Comment (fatcat-z) [inline onnxscript/tests/function_libs/torch_aten/ops_correctness_test.py:1026]: This is blocked by issue: https://github.com/microsoft/onnxruntime/issues/15446. Will update the code once ORT team confirms it is a bug.

**Comment by fatcat-z** [comment]
All failed tests are relative to aten::scaled_dot_product_attention op. Am I missing anything?

File: onnxscript/function_libs/torch_aten/ops/core.py [hunk 2]
```diff
@@ -3615,20 +3614,21 @@ def aten_max_pool2d(
     else:
         strides = stride
 
-    return _aten_max_pool2d_onnx(self, kernel_shape, strides, pads, dilations, ceil_mode)
+    return _aten_max_pool_onnx(self, kernel_shape, strides, pads, dilations, ceil_mode, 3)
 
 
-@torch_op("aten::max_pool2d", private=True)
-def _aten_max_pool2d_onnx(
+@torch_op("aten::max_pool", private=True)
```
Comment (titaiwangms) [inline onnxscript/function_libs/torch_aten/ops/core.py:3620]: @fatcat-z @justinchuby @xiaowuhu Although `aten::max_pool` seems to be a fine custom opset name here, it does't actually exist. Should we come up with a general solution/rule for shared ONNX Function?

Comment (titaiwangms) [reply onnxscript/function_libs/torch_aten/ops/core.py:3620]: Or private=True should already indicated that to users?

**Comment by justinchuby** [comment]
> All failed tests in ubuntu are relative to aten::scaled_dot_product_attention op. Am I missing anything?

No you didn't miss anything. It is https://github.com/microsoft/onnx-script/issues/567

File: onnxscript/function_libs/torch_aten/ops/core.py (hunk 2 above)
Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:3620]: I agree that it'd be better for us to register private functions to an actual aten name unless there is a huge benefit to share code. In this case I suggest duplicating the `_aten_max_pool2d_onnx` logic to a separeate `_aten_max_pool3d_onnx` for clarity.

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:3620]: I do see why we want to share logic in `_aten_max_pool_with_indices_onnx` though...

Comment (justinchuby) [reply onnxscript/function_libs/torch_aten/ops/core.py:3620]: So this name doesn't _really_ matter for private functions. We may be able to get away from this slight deviation for now.

**Review by justinchuby — APPROVED** [review]

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 3]
```diff
@@ -617,17 +664,103 @@ def aten_max_pool2d_with_indices_backward(
...
+    pool_result, indices = op.MaxPool(
+        self,
+        ceil_mode=ceil_mode,
+        dilations=dilation,
+        kernel_shape=kernel_size,
+        pads=padding,
+        strides=stride,
+    )
+
+    _, flatten_indices = op.MaxPool(
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/nn.py:748]: I would add a comment to explain the smart things done here using the second maxpool

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 4]
```diff
@@ -617,17 +664,103 @@ def aten_max_pool2d_with_indices_backward(
...
+        pads = padding
+
+    if isinstance(stride, int):
+        strides = [stride] * expand_size
+    elif stride is None:
+        strides = kernel_shape
+    else:
+        strides = stride
+
+    return _aten_max_pool_with_indices_onnx(
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/nn.py:708]: Note: examples

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 5]
```diff
@@ -617,17 +664,103 @@ def aten_max_pool2d_with_indices_backward(
...
+        self,
+        kernel_shape,
+        strides,
+        pads,
+        dilations,
+        ceil_mode,
+        4,
+        ([1] * expand_size),
+        ([0] * expand_size),
+        ([2 + i for i in range(expand_size)]),
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/nn.py:718]: Magic 🪄

File: onnxscript/function_libs/torch_aten/ops/nn.py [hunk 6]
```diff
@@ -617,17 +664,103 @@ def aten_max_pool2d_with_indices_backward(
...
+    )
+
+
+@torch_op("aten::max_pool_with_indices", private=True)
```
Comment (justinchuby) [inline onnxscript/function_libs/torch_aten/ops/nn.py:722]:
```suggestion
@torch_op("internal::max_pool_with_indices", private=True)
```

**Comment by fatcat-z** [comment]
Close this one because of a merge mess. Re-create #618 for this.

**Pull request closed without merging**